src/NHibernate.Test/CacheTest/QueryKeyFixture.cs

```diff
@@ -100,6 +100,20 @@ public void NotEqualHashCodeWithFilters()
 			Assert.That(qk1.GetHashCode(), Is.Not.EqualTo(qvk1.GetHashCode()), "qk1 & qvk1");
 		}
 
+		[Test]
+		public void HashCodeWithFiltersAndSerialization()
```

Comment:
This new test does not demonstrate any failure in master. A multi-process test using .Net Core is likely required for this, but that will be a bit complex to setup.

Reply:
But after the change it fails... The hashcode computation is not stable itself. We have another bug there.

Reply:
@fredericDelaporte  I think you figure it out, sorry for miss code but I am new to nhibernate and i want it to contribute too much :)

Did you fix it with deserialization call back?

Reply:
@fredericDelaporte  I think FilterKey has own deserialization call back too.

Reply:
Not yet, but it is going to have it, since I do not see any other solution.

Reply:
No it is unrelated to the fields being readonly. Binary serialization has no issue with this.

My last fix was still not good. Maybe we will have better to cease caching hascodes.

Reply:
Also I am thinking the distributed cache scenario, "Multiple application single distributed cache" I did not test it but if it rely on HashCode, I think that they will also fail and cache the same result again right? Or just cache the query?

So maybe as you said we did not rely HashCode, just check Equals.

Reply:
No, I did not mean this. It is not an option in my opinion. Equal objects must have equal hashcodes, at least within the process in which they are compared. Any other solution is a workaround not fixing the amin issue. And moreover, since hashcode are essential for fast lookups, any other solution will imply a slow lookup logic.